### DISCUSSION TITLE
[@types/node] Should async EventEmitter `on` handlers be allowed?

### DISCUSSION THREAD

**Discussion by slifty** [DefinitelyTyped/DefinitelyTyped#61265] [Issues with a @types package] [answered]
I'm using a NodeJS [EventEmitter](https://nodejs.org/api/events.html#class-eventemitter) and would like to use an async event handler (this is related to some network middleware, so everything about the event "stream" is inherently asynchronous).

The current [type definition](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a148cf1f1797d00acd7d572ba3bd569ee955404f/types/node/events.d.ts#L348) requires event handlers to return void:

```typescript
on(eventName: string | symbol, listener: (...args: any[]) => void): this;
```

The node spec doesn't seem to indicate async functions aren't allowed / are to be avoided as `on` handlers.

Would it be appropriate to expand the handler signature to include `Promise<void>` in order to allow handlers to be async?

e.g. something like

```typescript
on(eventName: string | symbol, listener: (...args: any[]) => void | Promise<void>): this;
```

[a comment by an automated account is not shown]

**Comment by mcollina** [comment, marked as answer]
Don't use async handlers for on(), you'll save yourself plenty of trouble at later times.

**Reply by slifty** [reply]
Understood -- is that a universal truth, including for EventEmitters that are designed for asynchronous handlers?  For instance, I am working in a situation where an SFTP event emitter implements EventEmitter and uses `on` to propagate SFTP commands that were received from over a network connection.  It passes a "response" / "ready" argument to the event handler which is expected to be called with the appropriate response once it is ready to send back to the client.

I do appreciate the answer might be / is probably still "yes!" -- that this kind of situation is such an edge case that it's better to signal that async doesn't belong in the land of events, (and that's why it's currently not allowed).  In those cases the current spec does require somewhat akward code such as:

```
myNetworkProtocolEventEmitter.on(
  'myAsyncEvent',
  (argument1, argument2) => { myAsyncNetworkEventHandler(argument1, argument2); },
);`
```

(thank you for the response by the way!)

----

*Update:* A friend pointed out to me that the reason the answer is, indeed, "yes" is because nothing about `EventEmitter` is set up to handle promise rejections from an async handler.  This means one *should* be required to make the asynchronous call from within in a synchronous handler (and that call, ideally, would explicitly address any potential promise rejections).
